common/nexus/payload_serializer.go

```diff
@@ -0,0 +1,185 @@
+// The MIT License
+//
+// Copyright (c) 2023 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nexus
+
+import (
+	"errors"
+	"fmt"
+	"maps"
+	"mime"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	commonpb "go.temporal.io/api/common/v1"
+)
+
+type payloadSerializer struct{}
+
+var errSerializer = errors.New("serializer error")
+
+// Deserialize implements nexus.Serializer.
+func (payloadSerializer) Deserialize(content *nexus.Content, v any) error {
+	payloadRef, ok := v.(**commonpb.Payload)
+	if !ok {
+		return fmt.Errorf("%w: cannot deserialize into %v", errSerializer, v)
+	}
+
+	payload := &commonpb.Payload{}
+	*payloadRef = payload
+	payload.Metadata = make(map[string][]byte)
+	payload.Data = content.Data
+
+	h := maps.Clone(content.Header)
+	// We assume that encoding is handled by the transport layer and the content is decoded.
+	delete(h, "encoding")
+	// Length can safely be ignored.
+	delete(h, "length")
+
+	if len(h) > 1 {
+		setUnknownNexusContent(h, payload.Metadata)
+		return nil
+	}
+
+	contentType := h.Get("type")
+	if contentType == "" {
+		if len(h) == 0 && len(content.Data) == 0 {
+			payload.Metadata["encoding"] = []byte("binary/null")
+		} else {
+			setUnknownNexusContent(h, payload.Metadata)
+		}
+		return nil
+	}
+
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		setUnknownNexusContent(h, payload.Metadata)
+		return nil
+	}
+
+	switch mediaType {
+	case "application/x-temporal-payload":
+		err := payload.Unmarshal(content.Data)
+		if err != nil {
+			return err
+		}
+	case "application/json":
+		if len(params) == 0 {
+			payload.Metadata["encoding"] = []byte("json/plain")
+		} else if len(params) == 2 && params["format"] == "protobuf" && params["message-type"] != "" {
```

Comment:
I think you're being overly strict on the mime type params. If it's our content type, we can ignore params we don't care about

Reply:
again, I'd rather not lose this information. Either you're sticking to the standard or you're a custom nexus payload.